### ISSUE TITLE
slow performance when opening a file with tramp

### ISSUE THREAD

**Issue by jave** [domtronn/spaceline-all-the-icons.el#34]
It seems related to s-a-t-i because I have no performance problems when using plain modeline, and tramp together.

I did some profiling. It seems some functions are called a lot. Shouldnt it be possible to improve performance with memoization for instance? Failing that, can I disable spaceline-all-the-icons--buffer-path  perhaps?

- redisplay_internal (C function)                               27726  56%
 - eval                                                         27723  56%
  - spaceline-ml-all-the-icons                                  27721  56%
   - spaceline-all-the-icons--buffer-path                       22453  45%
    - projectile-project-root                                   22404  45%
     - cl-some                                                  22232  45%
      + #<compiled 0x599ded3>                                   22232  45%
     + file-remote-p                                              172   0%
    + file-truename                                                30   0%
    + file-name-directory                                          12   0%
      replace-regexp-in-string                                      3   0%
      split-string                                                  2   0%
      getenv                                                        1   0%
   + projectile-project-p                                        5221  10%
   + powerline-raw                                                 14   0%
   + all-the-icons-icon-for-mode                                    8   0%
   + spaceline-all-the-icons--face-background                       7   0%
   + file-name-nondirectory                                         4   0%
   + spaceline-all-the-icons--highlight                             3   0%
   + spaceline-all-the-icons--flycheck-status                       2   0%
     all-the-icons-faicon                                           2   0%
     make-mode-line-mouse-map                                       1   0%
   + spaceline-all-the-icons--separator                             1   0%
 + #<compiled 0x447507>                                             2   0%
+ command-execute                                               12290  24%
+ ...                                                            8961  18%
+ timer-event-handler                                             200   0%
+ tramp-file-name-handler                                           1   0%

**Comment by jave** [comment]
I tried turning off segments to find a workable setting for tramp buffers, but I couldnt really find any.

also it seemed like projectile-project-p was called even if I turned off that segment.

So I made a little workaround. Revert to standard mode-line when slowness is perceived.
The workaround works well, so there is indeed a connection between tramp buffers, spaceline-all-the-icons and slowness.


  (defun jv-std-modeline ()
    (interactive)
    (setq mode-line-format '("%e" mode-line-front-space mode-line-mule-info mode-line-client mode-line-modified mode-line-remote mode-line-frame-identification mode-line-buffer-identification "   " mode-line-position
   (vc-mode vc-mode)
   "  " mode-line-modes mode-line-misc-info mode-line-end-spaces))        )

**Comment by domtronn** [comment]
Hi @jave, yep, this is something I've been working on when I get chance. The latest version of [`all-the-icons`](https://github.com/domtronn/all-the-icons.el) include memoization so that should hopefully improve things for you!

Basically, I noticed this because there was a recent change to use the files _filename_ to calculate the icon in `all-the-icons-mode-icon` segment, which performs a _tonne_ of regexp matches. I've also changed this to use the the files `major-mode` which is a quick alist assoc.

I don't really use tramp in my day to day, but I can imagine that my `buffer-path` mode doesn't like it at all, so I could maybe add something to the predicate to disable/change that?

Also, I think Spaceline runs the `:when` predicates _even when_ the segment itself is disabled, which is unfortunate... So it may mean that the best design pattern is to perform the predicates as part of the main body of the segment if that makes sense?

*tl;dr*

What version of `all-the-icons` are you on? Maybe try updating to get the memoization.
If you could tell me what kind of behaviour you'd like in a tramp buffer, I can start working on the compatibility with the `buffer-path` segment

**Comment by jave** [comment]
I use spaceline-all-the-icons-20170506.408 from melpa. So theres a newer one?

**Comment by domtronn** [comment]
Nope, that's the latest version, but the library that it depends on _(`all-the-icons`)_ needs to be `20170513.1019` to get the memoization optimisations.

Which is _newer_ than the latest `spaceline-all-the-icons`. It's worth just checking that dependency.

However, like you say, there'll probably still be issues with tramp and the `buffer-path` segment

**Comment by jave** [comment]
ok, thanks, that dependency was indeed a couple of days older!

**Comment by SavchenkoValeriy** [comment]
I don't know if this related or not, but Projectile mode-line has a similar problem (with sudo and TRAMP): https://github.com/bbatsov/prelude/issues/594
